libnopegl/src/pass.c

```diff
@@ -476,6 +483,56 @@ static int build_uniforms_map(struct pipeline_desc *desc, struct darray *crafter
     return 0;
 }
 
+static int build_blocks_map(struct pass *s, struct pipeline_desc *desc)
+{
+    ngli_darray_init(&desc->blocks_map, sizeof(struct resource_map), 0);
+
+    struct pipeline_layout layout = ngli_pgcraft_get_pipeline_layout(desc->crafter);
+
+    for (size_t i = 0; i < layout.nb_buffer_descs; i++) {
+        const struct pipeline_resource_desc *resource = &layout.buffer_descs[i];
+        const struct hmap *resources = NULL;
+        if (resource->stage == NGLI_PROGRAM_SHADER_VERT)
+            resources = s->params.vert_resources;
+        else if (resource->stage == NGLI_PROGRAM_SHADER_FRAG)
+            resources = s->params.frag_resources;
+        else if (resource->stage == NGLI_PROGRAM_SHADER_COMP)
+            resources = s->params.compute_resources;
+        else
+            ngli_assert(0);
+
+        if (!resources)
+            continue;
+
+        const char *name = ngli_pgcraft_get_symbol_name(desc->crafter, resource->id);
+        const int32_t index = ngli_pgcraft_get_buffer_index(desc->crafter, name, resource->stage);
+
+        /* Remove _block suffix that has been added by pgcraft */
+        char resource_name[MAX_ID_LEN];
+        snprintf(resource_name, sizeof(resource_name), "%s", name);
+        size_t len = strlen(name);
+        size_t suffix_len = strlen("_block");
+        resource_name[NGLI_CLAMP(len - suffix_len, 0, MAX_ID_LEN - 1)] = 0;
```

Comment:
Untested but I believe safer & simpler:

```c
const size_t len = strlen(name);
const size_t suffix_len = strlen("_block");
ngli_assert(len > suffix_len && !strcmp(name + len, "_block"));
snprintf(resource_name, sizeof(resource_name), "%*.s", (int)(len - suffix_len), name);
```

A note on your version: I don't think the clamp can work for the lower bound because you're working with unsigned (at least not all all architectures).

Reply:
Done and adjusted your suggestion:
```
ngli_assert(len > suffix_len && !strcmp(name + (len - suffix_len), "_block"));
```